[Misc] Respect trace headers in grpc server

ronensc opened the pull request:
This PR updates the gRPC server to include the trace context of an incoming request in the reported trace.
It's a follow up on https://github.com/vllm-project/vllm/pull/4687.


## How to test this PR:

1. Install OpenTelemetry packages:
    ```
    pip install \
        opentelemetry-sdk \
        opentelemetry-api \
        opentelemetry-exporter-otlp \
        opentelemetry-semantic-conventions-ai
    ```

1. Start Jaeger in a docker container:
    ```
    # From: https://www.jaegertracing.io/docs/1.57/getting-started/
    docker run --rm --name jaeger \
        -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
        -p 6831:6831/udp \
        -p 6832:6832/udp \
        -p 5778:5778 \
        -p 16686:16686 \
        -p 4317:4317 \
        -p 4318:4318 \
        -p 14250:14250 \
        -p 14268:14268 \
        -p 14269:14269 \
        -p 9411:9411 \
        jaegertracing/all-in-one:1.57
    ```
	
1. In a new shell, export Jaeger IP:
    ```
    export JAEGER_IP=$(docker inspect   --format '{{ .NetworkSettings.IPAddress }}' jaeger)
    export OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=grpc://$JAEGER_IP:4317
    ```
    Then set vLLM's service name for OpenTelemetry, enable insecure connections to Jaeger and run vLLM with the grpc endpoint:
    ```
    export OTEL_SERVICE_NAME="vllm-server"
    export OTEL_EXPORTER_OTLP_TRACES_INSECURE=true
    python -m vllm.entrypoints.openai.api_server --model="facebook/opt-125m" --otlp-traces-endpoint="$OTEL_EXPORTER_OTLP_TRACES_ENDPOINT" --grpc-port 50051
    ```
	
1. In a new shell, send requests with trace context from a dummy grpc client:
    ```
    export JAEGER_IP=$(docker inspect --format '{{ .NetworkSettings.IPAddress }}' jaeger)
    export OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=grpc://$JAEGER_IP:4317
    export OTEL_EXPORTER_OTLP_TRACES_INSECURE=true
    export OTEL_SERVICE_NAME="client-service"
    python examples/production_monitoring/dummy_client_grpc.py
    ```
	
1. Open Jaeger webui: http://localhost:16686/ and find two traces: one from the dummy client containing the prompt text and one from vLLM containing metadata about the request.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

maxdebayser commented on vllm/engine/llm_engine.py:
```diff
@@ -785,6 +807,9 @@ def step(self) -> List[Union[RequestOutput, EmbeddingRequestOutput]]:
         self.do_log_stats(scheduler_outputs, output)
 
+        # Tracing
+        self.do_tracing(scheduler_outputs)
```
I'm concerned that we could be inserting a blocking network call here in the middle of the event look. What if at this place the scheduler outputs are enqueued somewhere and then in the `async_llm_engine.run_engine_loop()` the actual creation and sending of spans is offloaded to another thread pool?

FYI, I tried to research this a little bit and while I didn't find a good solution, the discussion in this issue has an interesting discussion on the subject: https://github.com/open-telemetry/opentelemetry-python/issues/62

ronensc replied:
I find that one of the key principles of OpenTelemetry is:
```
Library should not block end-user application by default.
```
from: https://opentelemetry.io/docs/specs/otel/performance/



Additionally, it appears that the `BatchSpanProcessor` uses a worker thread to export traces, thereby avoiding blocking the user code.
https://github.com/open-telemetry/opentelemetry-python/blob/257e6fe3345d42203109128a6ccc9afbb4b777bc/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L189-L191

maxdebayser replied:
Maybe we could try a few runs with the fmperf tool just to be sure that there is no impact on the key metrics

ronensc replied:
I measured the e2e overhead of sending traces using `benchmark_latency.py` and reported the benchmark results in a comment on the upstream PR: https://github.com/vllm-project/vllm/pull/4687#issuecomment-2160194922

maxdebayser commented on benchmarks/benchmark_latency.py:
```diff
@@ -19,25 +19,28 @@ def main(args: argparse.Namespace):
+        block_size=args.block_size,
+        gpu_memory_utilization=args.gpu_memory_utilization,
+        distributed_executor_backend=args.distributed_executor_backend,
+        otlp_traces_endpoint=args.otlp_traces_endpoint,
```
Is the `vllm/entrypoints/llm.py` file missing from this PR? I couldn't find the diff adding the `otlp_traces_endpoint` parameter. 

But the LLM class is meant for offline inference. Is there a use case for adding tracing information to offline generation requests?

ronensc commented:
Rebased the PR once `main` has rebased to include https://github.com/vllm-project/vllm/pull/4687.

ronensc commented:
The failure in the `Build / build (pull_request)` check seems unrelated to this PR. It error occurs while building the UBI docker image:
https://github.com/IBM/vllm/actions/runs/9580069640/job/26413904119#step:10:3159
```
...
Dockerfile.ubi:294
--------------------
 293 |     # Install libsodium for Tensorizer encryption
 294 | >>> RUN --mount=type=bind,from=libsodium-builder,src=/usr/src/libsodium,target=/usr/src/libsodium \
 295 | >>>     cd /usr/src/libsodium \
 296 | >>>     && make install
 297 |     ENV LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
--------------------
ERROR: failed to solve: process "/bin/sh -c cd /usr/src/libsodium     && make install" did not complete successfully: exit code: 2
Error: buildx failed with: ERROR: failed to solve: process "/bin/sh -c cd /usr/src/libsodium     && make install" did not complete successfully: exit code: 2
```

Would a rerun make it pass?

joerunde commented:
> Would a rerun make it pass?

I hope so!

I'll look into this more though

ronensc commented:
I rebased once again and added OTel dependencies to the docker image. To make it work, I had to upgrade the `VLLM_WHEEL_VERSION` to the latest version and remove the pinning `grpcio-tools==1.63.0`. Otherwise, I encoutered the following error:


```
ERROR: Cannot install grpcio-tools==1.63.0 and opentelemetry-exporter-otlp-proto-grpc because these package versions have conflicting dependencies.
The conflict is caused by:
    grpcio-tools 1.63.0 depends on protobuf<6.0dev and >=5.26.1
    opentelemetry-proto 1.25.0 depends on protobuf<5.0 and >=3.19

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

joerunde reviewed (approved):
🌶️🌶️🌶️

tdoublep commented:
Closing as stale and this repo has been re-purposed for Spyre enablement.
